File: Binary Tree Postorder Traversal/Binary Tree Preorder Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<Integer> postorderTraversal(TreeNode root) {
        ArrayList<Integer> list=new ArrayList<Integer>();
        Stack<TreeNode> stack=new Stack<TreeNode>();
        TreeNode temp=root;
        TreeNode lastvisited=null;
        TreeNode peeknode;
        while(!stack.isEmpty()||temp!=null){
            if(temp!=null){
                stack.push(temp);
                temp=temp.left;
            }
            else{
                peeknode=stack.peek();
                if(peeknode.right!=null&&lastvisited!=peeknode.right){
                    temp=peeknode.right;
                }else{
                    stack.pop();
                    list.add(peeknode.val);
                    lastvisited=peeknode;
                }
            }
        }
        return list;
    }
}

File: Binary Tree Postorder Traversal/Binary Tree Preorder Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def postorderTraversal(self, root):
        stack=[]
        postorder=[]
        temp=root
        lastvisit=None
        while len(stack)!=0 or temp!=None:
            if temp!=None:
                stack.append(temp)
                temp=temp.left
            else:
                peeknode=stack[-1]
                if peeknode.right!=None and lastvisit!=peeknode.right:
                    temp=peeknode.right
                else:
                    stack.pop()
                    postorder.append(peeknode.val)
                    lastvisit=peeknode
        return postorder
